File: HW/HW2/HW2.py
firstBadVersion = None
def isBadVersion(version):
    return version < firstBadVersion

def findFirstBadVersion(versions):
    start, end = 0, len(versions)-1
    first = versions[-1]
    while start < end:
        mid = (start + end) / 2
        if isBadVersion(versions[mid]):
            first = versions[mid]
            end = mid
        else:
            start = mid + 1
    return first

File: HW/HW2/HW2.java
package HW.HW2;

public class HW2 {
    int firstBadVersion;
    boolean isBadVersion(int version) {
        return version < firstBadVersion;
    }

    public int findFirstBadVersion(int[] versions) {
        int start = 0, end = versions.length - 1;
        int first = versions[versions.length - 1];
        while (start <= end) {
            int mid = (start + end) / 2;
            if (isBadVersion(versions[mid])) {
                first = versions[mid];
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return first;
    }
}
